kcidb/monitor/spool/__init__.py

```diff
@@ -97,29 +96,46 @@ def post(self, notification, timestamp=None):
             timestamp:      An "aware" datetime.datetime object specifying the
                             notification creation time, or None to use
                             datetime.datetime.now(datetime.timezone.utc).
+            due:            An "aware" datetime.datetime object specifying the
+                            time notification should be sent out,
+                            or None to use "timestamp".
 
         Returns:
-            True, if the notification was posted onto the spool,
-            False, if not (it was already there).
+            True, if the new notification was posted onto the spool,
+            or the existing, "unpicked" notification was updated,
+            False, if not (it existed and was "picked" at the moment).
         """
         assert isinstance(notification, Notification)
         assert timestamp is None or \
             isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
         if timestamp is None:
             timestamp = datetime.datetime.now(datetime.timezone.utc)
+        assert due is None or \
+            isinstance(due, datetime.datetime) and due.tzinfo
+        if due is None:
+            due = timestamp
+        message = notification.render().as_string(policy=email.policy.SMTPUTF8)
+        doc = self._get_doc(notification.id)
 
-        try:
-            self._get_doc(notification.id).create(dict(
+        @firestore.transactional
+        def create_or_update(transaction, message, doc):
+            snapshot = doc.get(field_paths=["picked_until"],
+                               transaction=transaction)
+            picked_until = snapshot.get("picked_until")
+            # If the document doesn't exist, or
+            # the timestamp is passed the picked_until field
```

Comment:
The second part of the comment is a bit redundant and says basically what the condition says. I think it will be more helpful if we say what it actually means:
```suggestion
            # the notification is "picked"
```